lib/mayaUsd/ufe/UsdAttribute.cpp

```diff
@@ -1215,168 +800,178 @@ template <typename T> void TypedUsdAttribute<T>::set(const T& value)
 {
-    auto attr = std::make_shared<UsdAttributeDouble>(item, usdAttr);
+    auto attr = std::make_shared<UsdAttributeDouble>(item, attrHolder);
     return attr;
 }
 
 //------------------------------------------------------------------------------
 // UsdAttributeString:
 //------------------------------------------------------------------------------
 
-/*static*/
-#ifdef UFE_V4_FEATURES_AVAILABLE
-UsdAttributeString::Ptr UsdAttributeString::create(
-    const UsdSceneItem::Ptr&           item,
-    const PXR_NS::UsdPrim&             prim,
-    const Ufe::AttributeDef::ConstPtr& attrDef)
+UsdAttributeString::UsdAttributeString(
+    const UsdSceneItem::Ptr& item,
+    UsdAttributeHolder*      attrHolder)
+    : Ufe::AttributeString(item)
+    , UsdAttribute(attrHolder)
 {
-    auto attr = std::make_shared<UsdAttributeString>(item, prim, attrDef);
-    return attr;
 }
-#endif
 
 /*static*/
 UsdAttributeString::Ptr
-UsdAttributeString::create(const UsdSceneItem::Ptr& item, const PXR_NS::UsdAttribute& usdAttr)
+UsdAttributeString::create(const UsdSceneItem::Ptr& item, UsdAttributeHolder* attrHolder)
 {
-    auto attr = std::make_shared<UsdAttributeString>(item, usdAttr);
+    auto attr = std::make_shared<UsdAttributeString>(item, attrHolder);
     return attr;
 }
 
+std::string UsdAttributeString::get() const
+{
+    if (!hasValue())
+        return std::string();
+
+    PXR_NS::VtValue vt;
+    if (UsdAttribute::get(vt, getCurrentTime(sceneItem())) && vt.IsHolding<std::string>()) {
+        return vt.UncheckedGet<std::string>();
+    }
+
+    return std::string();
+}
+
+void UsdAttributeString::set(const std::string& value)
+{
+    // We need to figure out if the USDAttribute is holding a TfToken or string.
+    setUsdAttr<std::string>(*this, value);
+}
+
+Ufe::UndoableCommand::Ptr UsdAttributeString::setCmd(const std::string& value)
+{
+    auto self = std::dynamic_pointer_cast<UsdAttributeString>(shared_from_this());
+    if (!TF_VERIFY(self, kErrorMsgInvalidType))
+        return nullptr;
+
+    const std::string errMsg = isEditAllowedMsg();
+    if (!errMsg.empty()) {
+        MGlobal::displayError(errMsg.c_str());
+        return nullptr;
+    }
+
+    return std::make_shared<SetUndoableCommand<std::string, UsdAttributeString>>(self, value);
+}
+
 //------------------------------------------------------------------------------
-// UsdAttributeColorFloat3:
+// UsdAttributeToken:
 //------------------------------------------------------------------------------
 
-/*static*/
-#ifdef UFE_V4_FEATURES_AVAILABLE
-UsdAttributeColorFloat3::Ptr UsdAttributeColorFloat3::create(
-    const UsdSceneItem::Ptr&           item,
-    const PXR_NS::UsdPrim&             prim,
-    const Ufe::AttributeDef::ConstPtr& attrDef)
+UsdAttributeToken::UsdAttributeToken(const UsdSceneItem::Ptr& item, UsdAttributeHolder* attrHolder)
+    : Ufe::AttributeString(item)
+    , UsdAttribute(attrHolder)
```

Comment:
And one to handle non-enum TfToken data. Might be unused though. Not sure we have a USD TfToken that allows any string to be entered. Kept for symmetry.